### PR TITLE
Initial add of distributed model

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #1060
* __->__ #1059

Use `parallelize_module` in model.

Added files:

`model_dist.py`: a mirror of model.py with Tensor Parallelism baked in.
`dist_run.py`: toy example of how to run the model in distributed way.

Test:
`torchrun --nproc-per-node 2 dist_run.py`

### CODE REVIEW

File: build/model_dist.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        if (
+            self.max_seq_length >= max_seq_length
+            and self.max_batch_size >= max_batch_size
+        ):
+            return
+        head_dim = self.config.dim // self.config.n_heads
+        max_seq_length = find_multiple(max_seq_length, 8)
+        self.max_seq_length = max_seq_length
+        self.max_batch_size = max_batch_size
+        for b in self.layers:
```
Comment (lessw2020) [inline build/model_dist.py:70]: maybe block instead of b?  Single letter vars can become confusing imo.

File: build/model_dist.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+                max_batch_size, max_seq_length, self.config.n_local_heads, head_dim
+            )
+
+        freqs_cis = precompute_freqs_cis(
+            self.config.dim // self.config.n_heads,
+            self.config.block_size * 2,
+            self.config.rope_base,
+            use_scaled = self.config.use_scaled_rope,
+        )
+        self.register_buffer("freqs_cis", freqs_cis, persistent=True)
```
Comment (lessw2020) [inline build/model_dist.py:81]: curious as to why persistent=True?  These are currently generated by default so not sure if there is a reason to start saving these?

File: build/model_dist.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        causal_mask = torch.tril(
+            torch.ones(self.max_seq_length, self.max_seq_length, dtype=torch.bool)
+        )
+        self.register_buffer("causal_mask", causal_mask, persistent=True)
```
Comment (lessw2020) [inline build/model_dist.py:85]: same q as above re: should persistent =True?

File: build/model_dist.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    ) -> Tensor:
+        h = x + self.attention(self.attention_norm(x), freqs_cis, mask, input_pos)
+        out = h + self.feed_forward(self.ffn_norm(h))
+        return out
+
+
+class Attention(nn.Module):
+    def __init__(self, config: TransformerArgs):
+        super().__init__()
+        assert config.dim % config.n_heads == 0
```
Comment (lessw2020) [inline build/model_dist.py:142]: add message?  , "Model embedding dims must be evenly divisible by num_heads"?

File: build/model_dist.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+        wq = nn.Linear(config.dim, config.n_heads * config.head_dim, bias=False)
+        wk = nn.Linear(
+            config.dim, config.n_local_heads * config.head_dim, bias=False
+        )
+        wv = nn.Linear(
+            config.dim, config.n_local_heads * config.head_dim, bias=False
+        )
+        wo = nn.Linear(config.dim, config.dim, bias=False)
+
+        self.wq = parallelize_module(wq, device_mesh, Colwise)
```
Comment (lessw2020) [inline build/model_dist.py:156]: wondering if it's cleaner to just go ahead and merge the self.wx creation and the parallelize into single calls ala:
```
self.wq = parallelize_module(
                           nn.Linear(config.dim, config.n_heads * config.head_dim, bias=False,) 
                           device_mesh, Colwise
                           )
```

File: build/model_dist.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+        q = q.view(bsz, seqlen, -1, self.head_dim)
+        k = k.view(bsz, seqlen, -1, self.head_dim)
+        v = v.view(bsz, seqlen, -1, self.head_dim)
+
+        q = apply_rotary_emb(q, freqs_cis)
+        k = apply_rotary_emb(k, freqs_cis)
+
+        q, k, v = (x.transpose(1, 2) for x in (q, k, v))
+
+        #if self.kv_cache is not None:
```
Comment (lessw2020) [inline build/model_dist.py:233]: mark with a TODO for updating KV caching support?

File: build/model_dist.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+        w3 = nn.Linear(config.dim, config.hidden_dim, bias=False)
+        self.w1 = parallelize_module(w1, device_mesh, Colwise)
+        self.w2 = parallelize_module(w2, device_mesh, Rowwise)
+        self.w3 = parallelize_module(w3, device_mesh, Colwise)
+
+    def forward(self, x: Tensor) -> Tensor:
+        y: DTensor = self.w2(F.silu(self.w1(x)) * self.w3(x))
+        # y is a DTensor with Partial placement;
+        # we convert its placement to Replicate and convert it back to a regular
+        # Tensor. `full_tensor` is the API that does both.
```
Comment (lessw2020) [inline build/model_dist.py:261]: thank you for these clarifying comments, very helpful!

File: dist_run.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        model.setup_caches(1, 4096)
+
+print(model)
+
+# Distributed run
+input_ids = torch.randint(0, config.vocab_size, (1, 4096), device=device)
+input_pos = torch.arange(0, 4096, device=device)
+output = model(input_ids, input_pos)
+dist.destroy_process_group()
+print(f"Rank {rank} completes.")
```
Comment (lessw2020) [inline dist_run.py:35]: add a __name__ == "__main__": 
     main()
just to make it clear this is a runnable file rather than an importing module?

**Review by lessw2020 — APPROVED** [review]
great stuff! 
minor nit comments for consideration.

File: build/model_dist.py (hunk 1 above)
Comment (kwen2501) [reply build/model_dist.py:70]: coming from original model, shall we keep them in line?

File: build/model_dist.py (hunk 2 above)
Comment (kwen2501) [reply build/model_dist.py:81]: coming from original model too.

File: build/model_dist.py (hunk 3 above)
Comment (kwen2501) [reply build/model_dist.py:85]: coming from original model too

File: dist_run.py (hunk 8 above)
Comment (kwen2501) [reply dist_run.py:35]: Fixed

File: build/model_dist.py (hunk 6 above)
Comment (kwen2501) [reply build/model_dist.py:233]: Added.

**Pull request merged**